### PR TITLE
Automatic reversal of DimCoord bounds

### LINKED ISSUE #423: ```Coord.is_contiguous()``` assumes too much
The `is_contiguous` method assumes that the lower bound is in index `0`, but when you reverse a coordinate, it is usually in index `1`. Because of this:

```
>>> from iris.coords import DimCoord
>>> c = DimCoord(tuple(range(10)), 'latitude')
>>> c.is_contiguous()
False
>>> c.guess_bounds()
>>> c.is_contiguous()
True
>>> c[::-1].is_contiguous()
False
>>> c[::-1].bounds         
array([[ 8.5,  9.5],
       [ 7.5,  8.5],
       [ 6.5,  7.5],
       [ 5.5,  6.5],
       [ 4.5,  5.5],
       [ 3.5,  4.5],
       [ 2.5,  3.5],
       [ 1.5,  2.5],
       [ 0.5,  1.5],
       [-0.5,  0.5]])
```

### LINKED ISSUE #3249: Fix ordering of cf coordinate bounds
It was pointed out that chopping a cube into slices over a dimension and re-merging (which was needed modify each part individually) had the effect of reversing a dimension coordinate with descending values.

And such a reversal breaks the 'is_contiguous' property :

E.g. 
```
>>> co_upwards = DimCoord(points=[20, 30, 40], bounds=[[15, 25], [25, 35], [35, 45]])
>>> co_ upwards.is_contiguous()
True
>>>
>>> co_downwards = co_upwards[::-1]
>>> co_downwards.is_contiguous()
False
```

Because what ‘contiguous’ means is just "bounds[i, 1] == bounds[i+1, 0], for all i"

According to : [cf 1.7, section 7.1](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#cell-boundaries), 
There under "Bounds for 1-D coordinate variables" we have : 
> If adjacent intervals are contiguous... For example, if the intervals that contain grid points lat(i) and lat(i+1) are contiguous, then latbnd(i+1,0) = latbnd(i,1)

So, our implementation of 'is_contiguous' does seems to be in line with the CF conventions,
but our reversed (dimension) coordinate effectively has ***invalid*** bounds,

### LINKED ISSUE #3756: iris.util.reverse does not flip coordinate bounds correctly
I found an issue in [iris.util.reverse](https://scitools.org.uk/iris/docs/latest/iris/iris/util.html#iris.util.reverse):

```python
import iris
print(iris.__version__)
print()

lat_coord = iris.coords.DimCoord([10, 20], bounds=[[5, 15], [15, 25]],
                                 standard_name='latitude')
cube = iris.cube.Cube([0.0, 1.0], dim_coords_and_dims=[(lat_coord, 0)])
reversed_cube = iris.util.reverse(cube, lat_coord)

print("Old lat coord:", cube.coord('latitude'))
print("Reversed lat coord:", reversed_cube.coord('latitude'))
print()
print("Old lat contiguous:", cube.coord('latitude').is_contiguous())
print("Reversed lat contiguous:", reversed_cube.coord('latitude').is_contiguous())
```

gives

```
2.4.0

Old lat coord: DimCoord(array([10, 20]), bounds=array([[ 5, 15],
       [15, 25]]), standard_name='latitude', units=Unit('1'))
Reversed lat coord: DimCoord(array([20, 10]), bounds=array([[15, 25],
       [ 5, 15]]), standard_name='latitude', units=Unit('1'))

Old lat contiguous: True
Reversed lat contiguous: False
```

As you can see, the bounds of the coordinate have not been reversed correctly: Instead of `[[15, 25], [5, 15]]`, they should be `[[25, 15], [15, 5]]`. As a side effect, `coord.is_contiguous()` fails now (that's how I discovered this issue).

### LINKED ISSUE #4078: Using iris.util.reverse flips coordinate bounds
## 🐛 Bug Report

When using `iris.util.reverse` along an axis, the cube that is returned is reversed properly. But the coordinate along which it has been reversed, has the bounds flipped and they become noncontiguous. This can cause issues when doing area-related operations.
## How To Reproduce
Steps to reproduce the behaviour:

An example for the latitudes would be

1.  cube = iris.util.reverse(cube, (1,))
2. cube.coord(var_name='lat').bounds

This is the current behaviour when we want to reverse a cube with a decreasing latitude coordinate to make it increasing instead.

Latitudes in a cube before reversing:

```
DimCoord(array([ 90.        ,  80.52631579,  71.05263158,  61.57894737,
        52.10526316,  42.63157895,  33.15789474,  23.68421053,
        14.21052632,   4.73684211,  -4.73684211, -14.21052632,
       -23.68421053, -33.15789474, -42.63157895, -52.10526316,
       -61.57894737, -71.05263158, -80.52631579, -90.        ]), bounds=array([[ 9.00000000e+01,  8.52631579e+01],
       [ 8.52631579e+01,  7.57894737e+01],
       [ 7.57894737e+01,  6.63157895e+01],
       [ 6.63157895e+01,  5.68421053e+01],
       [ 5.68421053e+01,  4.73684211e+01],
       [ 4.73684211e+01,  3.78947368e+01],
       [ 3.78947368e+01,  2.84210526e+01],
       [ 2.84210526e+01,  1.89473684e+01],
       [ 1.89473684e+01,  9.47368421e+00],
       [ 9.47368421e+00,  7.10542736e-15],
       [ 7.10542736e-15, -9.47368421e+00],
       [-9.47368421e+00, -1.89473684e+01],
       [-1.89473684e+01, -2.84210526e+01],
       [-2.84210526e+01, -3.78947368e+01],
       [-3.78947368e+01, -4.73684211e+01],
       [-4.73684211e+01, -5.68421053e+01],
       [-5.68421053e+01, -6.63157895e+01],
       [-6.63157895e+01, -7.57894737e+01],
       [-7.57894737e+01, -8.52631579e+01],
       [-8.52631579e+01, -9.00000000e+01]]), standard_name='latitude', units=Unit('degrees_north'), long_name='Long name', var_name='lat')
```
Latitudes after reversing:

```
DimCoord(array([-90.        , -80.52631579, -71.05263158, -61.57894737,
       -52.10526316, -42.63157895, -33.15789474, -23.68421053,
       -14.21052632,  -4.73684211,   4.73684211,  14.21052632,
        23.68421053,  33.15789474,  42.63157895,  52.10526316,
        61.57894737,  71.05263158,  80.52631579,  90.        ]), bounds=array([[-8.52631579e+01, -9.00000000e+01],
       [-7.57894737e+01, -8.52631579e+01],
       [-6.63157895e+01, -7.57894737e+01],
       [-5.68421053e+01, -6.63157895e+01],
       [-4.73684211e+01, -5.68421053e+01],
       [-3.78947368e+01, -4.73684211e+01],
       [-2.84210526e+01, -3.78947368e+01],
       [-1.89473684e+01, -2.84210526e+01],
       [-9.47368421e+00, -1.89473684e+01],
       [ 7.10542736e-15, -9.47368421e+00],
       [ 9.47368421e+00,  7.10542736e-15],
       [ 1.89473684e+01,  9.47368421e+00],
       [ 2.84210526e+01,  1.89473684e+01],
       [ 3.78947368e+01,  2.84210526e+01],
       [ 4.73684211e+01,  3.78947368e+01],
       [ 5.68421053e+01,  4.73684211e+01],
       [ 6.63157895e+01,  5.68421053e+01],
       [ 7.57894737e+01,  6.63157895e+01],
       [ 8.52631579e+01,  7.57894737e+01],
       [ 9.00000000e+01,  8.52631579e+01]]), standard_name='latitude', units=Unit('degrees_north'), long_name='Long name', var_name='lat')
```

## Expected behaviour

The expected behaviour would be to have a coordinate that even though it has been reversed, the bounds are still contiguous:

```
DimCoord(array([-90.        , -80.52631579, -71.05263158, -61.57894737,
       -52.10526316, -42.63157895, -33.15789474, -23.68421053,
       -14.21052632,  -4.73684211,   4.73684211,  14.21052632,
        23.68421053,  33.15789474,  42.63157895,  52.10526316,
        61.57894737,  71.05263158,  80.52631579,  90.        ]), bounds=array([[-9.00000000e+01, -8.52631579e+01],
       [-8.52631579e+01, -7.57894737e+01],
       [-7.57894737e+01, -6.63157895e+01],
       [-6.63157895e+01, -5.68421053e+01],
       [-5.68421053e+01, -4.73684211e+01],
       [-4.73684211e+01, -3.78947368e+01],
       [-3.78947368e+01, -2.84210526e+01],
       [-2.84210526e+01, -1.89473684e+01],
       [-1.89473684e+01, -9.47368421e+00],
       [-9.47368421e+00,  7.10542736e-15],
       [ 7.10542736e-15,  9.47368421e+00],
       [ 9.47368421e+00,  1.89473684e+01],
       [ 1.89473684e+01,  2.84210526e+01],
       [ 2.84210526e+01,  3.78947368e+01],
       [ 3.78947368e+01,  4.73684211e+01],
       [ 4.73684211e+01,  5.68421053e+01],
       [ 5.68421053e+01,  6.63157895e+01],
       [ 6.63157895e+01,  7.57894737e+01],
       [ 7.57894737e+01,  8.52631579e+01],
       [ 8.52631579e+01,  9.00000000e+01]]), standard_name='latitude', units=Unit('degrees_north'), long_name='Long name', var_name='lat')
```

## Environment 
 - OS & Version: [e.g., Ubuntu 20.04 LTS]
 - Iris Version: '3.0.1'

## Additional context

<details>
<summary>Click to expand <b>this section...</b></summary>

Related to https://github.com/ESMValGroup/ESMValCore/issues/1060

```
Add additional verbose information in a collapsible section.
```
See [here](https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab) for further details.
</details>

### PR DESCRIPTION
**Pull request by rcomer**
### Description

Fixes #3249, following @pp-mo's suggestion at https://github.com/SciTools/iris/issues/3249#issuecomment-452417442 to reorder the bounds for each cell when we create a `DimCoord`, so that the bounds order is consistent with the coordinate's direction.  This is consistent with the _Bounds for 1-D coordinate variables_ subsection of the [Cell Boundaries](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.9/cf-conventions.html#cell-boundaries) section in the CF-Conventions.  It means that operations such as `my_dimcoord[::-1]` will preserve contiguity.

Also fixes #4078 and fixes #3756 (which are direct duplicates of each other) and fixes #423.

This change caused a test failure at [test_cube_extract_by_coord_advanced](https://github.com/SciTools/iris/blob/3638874a63e9bbee72b062fe02a8e500982eb796/lib/iris/tests/test_cdm.py#L812) because `subset` did not find any matches.  So I also modified cell equality to accept the bounds in either order: this seems intuitive to me if the order of bounds can silently reverse.

I note that #3249, #3756 and #4078 are all in the v3.3 milestone, but #423 is labelled for a major release.  **Should this change be controlled by a `FUTURE` flag**?  It does seem to meet the criteria of a [change in default behaviour](https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_deprecations.html#changing-a-default).

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

File: lib/iris/tests/integration/merge/test_merge.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        coord1 = DimCoord([3, 2, 1], long_name="spam")
+        coord1.guess_bounds()
+        cube1.add_dim_coord(coord1, 0)
+        cubes = CubeList(cube1.slices_over("spam"))
+        cube2 = cubes.merge_cube()
+        coord2 = cube2.coord("spam")
+
+        self.assertTrue(coord2.is_contiguous())
+        self.assertArrayEqual(coord2.points, [1, 2, 3])
+        self.assertArrayEqual(coord2.bounds, coord1.bounds[::-1, ::-1])
```
Comment (rcomer) [inline lib/iris/tests/integration/merge/test_merge.py:33]: This test covers the original problem raised at #3249.  I've not done anything in the integrations tests before and am very not sure that I know what I'm doing!

File: lib/iris/tests/unit/util/test_reverse.py [hunk 2]
```diff
@@ -135,6 +140,9 @@ def test_cube_coord(self):
         self.assertArrayEqual(
             self.cube2.coord("a").points, cube1_reverse0.coord("a").points
         )
+        self.assertArrayEqual(
+            self.cube2.coord("a").bounds, cube1_reverse0.coord("a").bounds
+        )
```
Comment (rcomer) [inline lib/iris/tests/unit/util/test_reverse.py:145]: Modified to cover #4078 and #3936.

**Review by wjbenfold — CHANGES_REQUESTED** [review]
Looking good :) Just a few comments

**Review by rcomer — COMMENTED** [review]
Thanks for another excellent review @wjbenfold!  I have addressed your comments and also added a whatsnew, which renders like this:

image

Found it a bit hard to describe this change, so suggestions on the wording are welcome!

**Comment by rcomer** [comment]
I've no idea what caused the pre-commit failure.  Re-spinning didn't help 🙁

**Comment by jamesp** [comment]
I think there has just been an API breaking change in `click`. we had a similar failure in another library.

**Comment by wjbenfold** [comment]
Looks like the latest `black` has it fixed, so maybe we want to refresh the lockfiles manually?

**Comment by rcomer** [comment]
See #4668 🙂

File: lib/iris/tests/unit/util/test_reverse.py [hunk 3]
```diff
@@ -94,76 +99,108 @@ def setUp(self):
...
         )
+        self.assertArrayEqual(
+            self.cube2.coord("a").bounds, cube1_reverse_both.coord("a").bounds
+        )
         self.assertArrayEqual(
             self.cube2.coord("b").points, cube1_reverse_both.coord("b").points
         )
 
-    def test_cube_coord(self):
+    def test_cube_coord0(self):
```
Comment (wjbenfold) [inline lib/iris/tests/unit/util/test_reverse.py:146]: I'm not super keen that we've repeated some code from `test_cube_dim0` identically here (c.f. to rest of file) - particularly as changing one instance would require that the other changed. The assertions could jump out to an auxiliary method that's called by both tests?

**Review by wjbenfold — CHANGES_REQUESTED** [review]
Looking good, just some small suggestions :)

File: lib/iris/tests/unit/util/test_reverse.py (hunk 3 above)
Comment (rcomer) [reply lib/iris/tests/unit/util/test_reverse.py:150]: how's this? https://github.com/SciTools/iris/pull/4466/commits/0dea5af6078d04209e35630e9dde7d55910eaae1

File: lib/iris/tests/unit/util/test_reverse.py [hunk 4]
```diff
@@ -94,85 +99,92 @@ def setUp(self):
...
+
+    def check_coordb_unchanged(self, result):
         self.assertArrayEqual(
-            self.cube2.coord("a").points, cube1_reverse_both.coord("a").points
+            self.cube1.coord("b").points, result.coord("b").points
         )
+
+    def test_cube_dim0(self):
+        cube1_reverse0 = reverse(self.cube1, 0)
+        self.assertArrayEqual(self.cube1.data[::-1], cube1_reverse0.data)
```
Comment (wjbenfold) [inline lib/iris/tests/unit/util/test_reverse.py:130]: You could add a line break here if you liked for consistency

**Review by wjbenfold — APPROVED** [review]
One comment of pretty much the smallest type. I like the rearrangement :)

**Pull request merged**